### DISCUSSION TITLE
How to automate the RunReportJob programmatically

### DISCUSSION THREAD

**Discussion by rakeshSgr** [Sunbird-Obsrv/Community#80] [Query]
Hi,

As part of the 6.0 release, we are automating the creation of admin reports and program dashboard reports using a Python script. In the script, we make API calls to create backend and frontend configurations.

After running the script, in order to make the reports available, we need to execute the RunReportJob service in Jenkins. This service, accessible at http://10.20.0.14:8080/jenkins/job/Deploy/job/dev/job/DataPipeline/job/RunReportJob/, is responsible for executing the Druid query associated with the reports. It retrieves the query results and converts them into CSV and JSON formats. These output files are then pushed to the cloud storage.

Running the RunReportJob service ensures that the latest data is fetched and made available for generating the reports. Currently, we need to manually run this job for each report.

@manjudr @kumarks1122 @santhosh-tg @tsprasath
When we create a new report, we need to manually run it via Jenkins to create a cron entry.
Please let us know, how we do it if not via Jenkins

CC : @aks30 @vijiurs @adithyadinesh0412

**Comment by manjudr** [comment]
@rakeshSgr 

There are multiple ways to automate it:

1. You can choose to modify your Jenkins jobs scheduler section to run the job once a day, according to your preferred frequency. This configuration will ensure that the Jenkins job is triggered on a daily basis. As an example, you can refer to the ETL Jobs and the Content Cache Updater Jenkins jobs.

2. Another approach is to create a simple Ansible role that configures the cron job within the machine. This job should only be deployed during the provisioning or deployment time.

**Reply by rakeshSgr** [reply]
@manjudr Could you please let us know a convenient time for us to connect? It would be beneficial as we can discuss the problem we are facing and work towards finding a solution together.

**Comment by rakeshSgr** [comment]
@manjudr @anandp504 
In release 6.0, we are faced with the task of creating over 60 charts ( around 10 reports). Currently, we have to run the RunReportJob 60 times individually to generate the necessary data. This process is time-consuming and inefficient.

To verify, We have created the report called "Test Report for Auto Update" in staging on July 7th, till date report data has not been generated.  
![Screenshot from 2023-07-11 17-11-46](https://github.com/Sunbird-Obsrv/Community/assets/49852443/c007d0ff-38b7-431e-a274-cd0e6c3c1dca)

We kindly request your assistance in resolving this issue. We are seeking a more streamlined and efficient solution to create all the required reports in a more automated and consolidated manner.


@aks30 @vijiurs @adithyadinesh0412

**Reply by manjudr** [reply]
@rakeshSgr - The above suggestions are not answering your queries.? - https://github.com/orgs/Sunbird-Obsrv/discussions/80#discussioncomment-6370590. Generally, we need to run all these report jobs every day right? if yes, then the above suggestions should help

**Reply by rakeshSgr** [reply]
It won't solve our problem, could you suggest a good time to connect? We can talk about this.

**Reply by manjudr** [reply]
@rakeshSgr  - No worries, we can connect. Could you please provide more details on why the above suggestions won't be effective in resolving the issue? What happens if you schedule them as cron jobs?

**Reply by rakeshSgr** [reply]
@manjudr 
If it is possible to create cron jobs from the script for the report ID, could you please share a sample or any existing implementation with us?

**Reply by manjudr** [reply]
@rakeshSgr 
Can we create an Ansible script to set up cron jobs for each report identifier? So, whenever a new report is introduced, we just need to trigger the corn scheduler job.

**Reply by manjudr** [reply]
Is this issue resolved? are you able to write an ansible script to set up the corn job?

**Reply by rakeshSgr** [reply]
@manjudr We don't know, how to set up this, Can you help us with this?  Can we connect for the same?

**Reply by manjudr** [reply]
Please refer to this code to setup the corn job through ansible - https://github.com/Sunbird-Obsrv/sunbird-data-pipeline/blob/master/ansible/roles/data-products-deploy/tasks/main.yml#L194-L215

**Reply by rakeshSgr** [reply]
@manjudr As per my understanding, we need to create a cron job for each reportId, and for each release, we may create multiple reports, so that the ansible script will keep growing, is this the correct implementation?
Please provide ur availability, so we can have a call to understand better.

**Reply by manjudr** [reply]
How many reports do we have @rakeshSgr I think deploying through Ansible should not be an issue.

**Reply by rakeshSgr** [reply]
In release 6.0, we are faced with the task of creating over 60 charts

**Reply by manjudr** [reply]
@rakeshSgr  I hope I have answered your queries over the call. 

Summary -  when we deal with a large volume of reports that need to be added to a registered, you can create a basic script using any programming language (python, javascript). This script can utilize the report APIs to generate requests and register the reports in the table. The API specifications are provided below. Please use this information to craft a script for report creation.

http://docs.sunbird.org/latest/apis/reports/#operation/ReadReport

**Reply by rakeshSgr** [reply]
Hi @manjudr 
Reports and charts data not getting generated after making Publish API call also. 
We need help in this @kumarks1122, please check this

**Reply by manjudr** [reply]
@rakeshSgr  - did you check the status of the request in the postgres table? What is the request status?

**Reply by rakeshSgr** [reply]
@manjudr Status of the report is live, but the data is not updated. 
What could be the issue?

**Reply by manjudr** [reply]
@rakeshSgr  - Do we have data in the druid for the configured data range? Is the data source available 100%? Is the report job running every day?

**Reply by rakeshSgr** [reply]
@manjudr We replicated the existing report and tried this. But the report job running is disabled in the staging.
http://docs.sunbird.org/latest/apis/reports/#operation/ReadReport This is not accessible now

**Reply by manjudr** [reply]
@rakeshSgr  - Please Sync up with the DevOps team if the jobs are required to enable.

**Reply by rakeshSgr** [reply]
@santhosh-tg and @tsprasath Please enable the report job in staging, so that we confirm its working or not.
cc : @manjudr @aks30

**Reply by tsprasath** [reply]
@rakeshSgr http://10.20.0.14:8080/jenkins/job/Deploy/job/staging/job/DataPipeline/job/RunReportjob/

**Reply by rakeshSgr** [reply]
@manjudr is the same job you are referring to? 
@tsprasath is this job run every day?

**Reply by manjudr** [reply]
Yes @rakeshSgr

**Reply by rakeshSgr** [reply]
@tsprasath is the job run every day?
